roast/hash.go

```diff
@@ -5,22 +5,57 @@ import (
 	"math/big"
 )
 
-// The contextString as defined in section 6.5. FROST(secp256k1, SHA-256) of
-// [FROST] is "FROST-secp256k1-SHA256-v1". Since we do a BIP-340 specialized
-// version, we use "FROST-secp256k1-BIP340-v1" as the contextString.
-var contextString = []byte("FROST-secp256k1-BIP340-v1")
+// Hash interface abstracts out hash functions implementations specific to the
+// ciphersuite used. This is a strategy design pattern allowing to use FROST
+// with different ciphersuites, like secp256k1 or Jubjub curves.
+//
+// [FROST] requires the use of a cryptographically secure hash function,
+// generically written as H. Using H, [FROST] introduces distinct domain-separated
+// hashes, H1, H2, H3, H4, and H5. The details of H1, H2, H3, H4, and H5 vary
+// based on ciphersuite.
+type Hash interface {
+	H1(m []byte) *big.Int
+	H2(m []byte, ms ...[]byte) *big.Int
```

Comment:
In the interface, we expect at least one message to be provided. Optionally, we allow for more and they will all be concatenated. Expecting at least one message makes sense for the hash function signature. We allow for more to be provided in the second argument because it is how this function will be used in FROST.

Reply:
Because `H2` has to conform to the specification of whatever FROST variant is being implemented, it needs special consideration. Different protocols have different rules for how to encode points in the signature and its verification. We can use multiple `[]byte`s as arguments if we use another interface (e.g. `Curve`) to define the encoding rules. Otherwise it'll need to be `H2(gc *Point, pk *Point, msg []byte) *big.Int` or something similar.

Reply:
I was thinking about having a `Cyphersuite` interface we would pass to the FROST protocol. The `Cyphersuite` would specify the particular `Hash`ing strategy but also the `Curve` where we could put the encoding logic.

Reply:
I think something along those lines makes sense.

Reply:
We could also rename `Hash` interface to `Cyphersuite`. We would keep `H*` and curve operations there. I am not sure which option is better, I'd decide about it later once we have more information.